### PR TITLE
Generalize mode-overlap detectors and S-parameter calculation to custom reference modes

### PR DESCRIPTION
**Pull request by bruxillensis**
Summary

ModeOverlapDetector currently only supports reference modes produced by the tidy3d waveguide mode solver. My motivation was that I wanted to use ModeOverlapDetector for gaussians and custom (user defined or solved for) modes, for example with inverse design of a grating into a radiating mode. This PR splits the overlap machinery into an abstract base class and adds two new detectors that overlap the simulated fields against user-supplied or analytic reference modes — enabling coupling-efficiency and S-parameter workflows for free-space beams or modes imported from other tools, without involving the mode solver.

What changed
  - BaseModeOverlapDetector (new ABC): owns everything independent of how the reference mode is produced — the stored mode fields, detector-plane face-area weights, the dispersion/conductivity correction (the reference mode sees the full complex ε(ω_c), same as before), and the overlap integral (compute_overlap / compute_overlap_to_mode). Subclasses implement a single hook, _compute_mode_fields, called once per wave character.
  - ModeOverlapDetector: unchanged behavior, now a subclass that implements the hook via compute_mode (tidy3d).
  - CustomModeOverlapDetector (new): takes a mode_function callable, invoked per frequency with the detector-plane coordinate meshgrids, the frequency, the propagation axis, and the effective inverse permittivity on the plane. Returns (mode_E, mode_H) in FDTDX's η₀-normalized convention; optional renormalization to unit Poynting flux (matching the mode solver).
  - GaussianModeOverlapDetector (new) + gaussian_mode_function: ready-made analytic Gaussian reference mode. Polarization and propagation tilt are configured exactly like GaussianPlaneSource (polarization axis or explicit vectors, azimuth_angle/elevation_angle in degrees), with the medium index derived from the permittivity slice — dispersion-aware via the shared base-class correction. Under a tilt, a transverse phase ramp matches the tilted wavefront; at normal incidence the mode stays real.
  - S-parameters: calculate_sparam / the sparams utilities accept any BaseModeOverlapDetector, so S-parameters can be computed against custom reference modes.
  - Shared polarization helper: the tilted E/H/k derivation used by the plane sources is factored into core.misc.tilted_polarization_vectors and reused by the analytic Gaussian mode, so source and detector stay on one convention by construction.

Usage
```
  det = fdtdx.GaussianModeOverlapDetector(
      name="fiber_coupling",
      partial_grid_shape=(None, None, 1),
      wave_characters=(fdtdx.WaveCharacter(wavelength=1.55e-6),),
      mode_radius=2.0e-6,
      direction="+",
  )

  # or an arbitrary reference mode:
  det = fdtdx.CustomModeOverlapDetector(
      name="custom",
      partial_grid_shape=(None, None, 1),
      wave_characters=(wave,),
      mode_function=my_mode,  # (coordinates, frequency, propagation_axis, inv_permittivity) -> (E, H)
  )
```

No behavior change for existing ModeOverlapDetector users; the refactor is purely structural and all public APIs are preserved.

## Summary by CodeRabbit

* **New Features**
  * Added custom and analytic Gaussian mode-overlap detectors.
  * Added support for tilted, polarized, and diverging Gaussian reference modes.
  * Exposed new mode-detector tools and helpers through the top-level API.
  * Added support for custom mode functions and optional flux normalization.
  * Extended S-parameter calculations to support all mode-overlap detector types.

* **Documentation**
  * Added the new mode-overlap detectors and Gaussian helpers to the API reference.

* **Bug Fixes**
  * Improved polarization and wave-vector handling for angled plane-wave sources.

* **Tests**
  * Added comprehensive validation for Gaussian modes, custom modes, dispersion, tilts, divergence, and detector overlap behavior.

### CODE REVIEW

**Comment by ymahlau** [comment]
Hi @bruxillensis, thanks for another PR! I skimmed over the code and the structural changes, which all look/sound good to me. If you want to mark the PR as ready for review I am happy to give a more detailed look at the code.

**Comment by bruxillensis** [comment]
Marked! Its exciting to see the contributor growth, you have many PR's to review now!

File: src/fdtdx/objects/detectors/mode.py [hunk 1]
```diff
@@ -285,3 +417,280 @@ def compute_overlap(
...
+                mode_H,
+                axis=self.propagation_axis,
+                area_weights=self._face_area_weights(),
+            )
+        return mode_E, mode_H, mode_neff
+
+
+@autoinit
+class GaussianModeOverlapDetector(BaseModeOverlapDetector):
+    """Mode-overlap detector using an analytic Gaussian beam as the reference mode.
```
Comment (ymahlau) [inline src/fdtdx/objects/detectors/mode.py]: I would maybe say "Tilted Gaussian" here not "Gaussian Beam" to avoid confusion with focused Gaussian beam, or generally disambiguate this difference a bit in the docstrings

File: src/fdtdx/objects/detectors/mode.py [hunk 2]
```diff
@@ -6,60 +8,181 @@
...
+            polarization_axis=polarization_axis,
+            fixed_E_polarization_vector=fixed_E_polarization_vector,
+            fixed_H_polarization_vector=fixed_H_polarization_vector,
+            azimuth_angle=azimuth_angle,
+            elevation_angle=elevation_angle,
+            center=center,
+            refractive_index=refractive_index,
+            wavenumber=wavenumber,
+            dtype=inv_permittivity.dtype,
+        )
```
[the remark this answers is not in the document]
Comment (ymahlau) [reply src/fdtdx/objects/detectors/mode.py:187]: @bruxillensis I think this does not need to be implemented, but the approximation should be documented

File: src/fdtdx/objects/detectors/mode.py [hunk 3]
```diff
@@ -6,60 +8,181 @@
...
+
+    t0, t1 = get_transverse_axes(propagation_axis)
+    transverse_0 = coordinates[t0] - center[0]
+    transverse_1 = coordinates[t1] - center[1]
+    amplitude = jnp.exp(-(transverse_0**2 + transverse_1**2) / (radius**2))
+
+    if is_tilted:
+        # Transverse phase ramp matching a tilted wavefront (phase 0 at the plane center).
+        phase_arg = wavenumber * (wave_vector[t0] * transverse_0 + wave_vector[t1] * transverse_1)
+        amplitude = amplitude * jnp.exp(1j * phase_arg)
```
[the remark this answers is not in the document]
Comment (ymahlau) [reply src/fdtdx/objects/detectors/mode.py:137]: @bruxillensis same here, either implement it or document the approximation. I think for the motivating example with gratings, this might actually be a problem, so it might be worthwhile to implement

**Comment by ymahlau** [comment]
I gave the code a closer look, one more comment in addition to the ones above:
- I think the test cases could be more extensive: A simulation test measuring overlap between Gaussian Source and OverlapDetector would validate the correct functionality

Other than that, the code looks good to me!

**Comment by bruxillensis** [comment]
@ymahlau I think this is ready for another look. Was on a work trip for the past few weeks. Comments were addressed, and I added divergence_angle since the original implementation assumed a perfectly collimated beam. It feels slightly unintuitive to me still to construct a gaussian detector (as in, you need to understand how the parameters are used), but maybe thats unavoidable. Open to feedback on it.

**Comment by ymahlau** [comment]
I also can't think of a better solution from the top of my head, but I think it is also okay to expect a little understanding of what the parameters do. Thanks for the PR!
